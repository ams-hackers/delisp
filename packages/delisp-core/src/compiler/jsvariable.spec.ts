import {
  escapeIdentifier,
  isValidJSIdentifier,
  isValidJSIdentifierName
} from "./jsvariable";

describe("Check valid JS identifier name", () => {
  it("should return false for invalid JS identifier names", () => {
    expect(isValidJSIdentifierName("foo?")).toBe(false);
    expect(isValidJSIdentifierName("123")).toBe(false);
    expect(isValidJSIdentifierName("2dVector")).toBe(false);
  });
  it("should return true for valid JS identifier names", () => {
    expect(isValidJSIdentifierName("isFoo")).toBe(true);
    expect(isValidJSIdentifierName("$123")).toBe(true);
    expect(isValidJSIdentifierName("_123")).toBe(true);
    expect(isValidJSIdentifierName("Vector2D")).toBe(true);
    expect(isValidJSIdentifierName("default")).toBe(true);
    expect(isValidJSIdentifierName("const")).toBe(true);
  });
});

describe("Check valid JS identifier", () => {
  it("should return false for invalid JS identifiers", () => {
    expect(isValidJSIdentifier("const")).toBe(false);
    expect(isValidJSIdentifier("default")).toBe(false);
  });
  it("should return true for valid JS identifiers", () => {
    expect(isValidJSIdentifier("_const")).toBe(true);
    expect(isValidJSIdentifier("$default")).toBe(true);
  });
});

describe("Variable name translation", () => {
  it("should escape delisp variable characters not allowed in Javascript", () => {
    const cases = [
      "$",
      "+",
      "-",
      "*",
      "/",
      "int?",
      "a->b",
      "<=",
      "%compile",
      "!foo",
      "&foo"
    ];
    const result = cases.map(v => ({
      delisp: v,
      js: escapeIdentifier(v)
    }));
    return expect(result).toMatchSnapshot();
  });

  it("should fail for not allowed characters", () => {
    expect(() => escapeIdentifier("``")).toThrow();
  });
});
