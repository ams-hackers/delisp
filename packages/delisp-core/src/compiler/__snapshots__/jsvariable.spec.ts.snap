// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Variable name translation should escape delisp variable characters not allowed in Javascript 1`] = `
Array [
  Object {
    "delisp": "$",
    "js": "$$dollar$",
  },
  Object {
    "delisp": "+",
    "js": "$$plus$",
  },
  Object {
    "delisp": "-",
    "js": "$$minus$",
  },
  Object {
    "delisp": "*",
    "js": "$$star$",
  },
  Object {
    "delisp": "/",
    "js": "$$div$",
  },
  Object {
    "delisp": "int?",
    "js": "$int$q$",
  },
  Object {
    "delisp": "a->b",
    "js": "$a$minus$$gt$b",
  },
  Object {
    "delisp": "<=",
    "js": "$$lt$$eq$",
  },
  Object {
    "delisp": "%compile",
    "js": "$$percent$compile",
  },
  Object {
    "delisp": "!foo",
    "js": "$$bang$foo",
  },
  Object {
    "delisp": "&foo",
    "js": "$$amp$foo",
  },
]
`;
